### PR TITLE
port the self-insert-command

### LINKED ISSUE #652: Port self-insert-command
When this is complete, cmds.c will be empty.

### PR DESCRIPTION
**Pull request by purplejacket**
but not internal_self_insert (yet)
this fixes #652 
I'm planning to port function internal_self_insert of cmds.c next

### CODE REVIEW

**Comment by shaleh** [comment]
I started this but did not get a chance to debug it....

**Comment by shaleh** [comment]
```
diff --git a/rust_src/remacs-sys/lib.rs b/rust_src/remacs-sys/lib.rs
index c75fd8f..babd1bd 100755
--- a/rust_src/remacs-sys/lib.rs
+++ b/rust_src/remacs-sys/lib.rs
@@ -1697,6 +1697,10 @@ extern "C" {
     pub fn del_range(from: ptrdiff_t, to: ptrdiff_t);
     pub fn buf_bytepos_to_charpos(b: *mut Lisp_Buffer, bytepos: ptrdiff_t) -> ptrdiff_t;
     pub fn swap_in_symval_forwarding(sym: *mut Lisp_Symbol, blv: *mut Lisp_Buffer_Local_Value);
+    pub fn internal_self_insert(c: c_int, n: EmacsInt) -> c_int;
+    pub fn bitch_at_user() -> !;
+    pub fn translate_char(table: Lisp_Object, c: c_int) -> c_int;
+    pub fn frame_make_pointer_invisible(f: *mut Lisp_Frame);
 
 }
 
diff --git a/rust_src/src/cmds.rs b/rust_src/src/cmds.rs
index 6aec6ad..c1bbaac 100644
--- a/rust_src/src/cmds.rs
+++ b/rust_src/src/cmds.rs
@@ -2,13 +2,19 @@
 
 use std::ffi::CString;
 
+use libc;
+
 use remacs_macros::lisp_fn;
 use remacs_sys::{Qbeginning_of_buffer, Qend_of_buffer, Qkill_forward_chars, Qnil,
-                 Qundo_auto_amalgamate};
-use remacs_sys::{del_range, initial_define_key, scan_newline_from_point, set_point, set_point_both};
+                 Qundo_auto__this_command_amalgamating, Qundo_auto_amalgamate};
+use remacs_sys::{bitch_at_user, del_range, frame_make_pointer_invisible, globals,
+                 initial_define_key, internal_self_insert, scan_newline_from_point, set_point,
+                 set_point_both, translate_char};
 use remacs_sys::EmacsInt;
+use remacs_sys::Fset;
 
 use editfns::{line_beginning_position, line_end_position};
+use frames::selected_frame;
 use keymap::{current_global_map, Ctl};
 use lisp::LispObject;
 use lisp::defsubr;
@@ -185,18 +191,62 @@ pub fn forward_line(n: Option<EmacsInt>) -> EmacsInt {
     }
 }
 
-pub fn initial_keys() {
+// Note that there's code in command_loop_1 which typically avoids
+// calling this.
+/// Insert the character you type.
+/// Whichever character you type to run this command is inserted.
+/// The numeric prefix argument N says how many times to repeat the insertion.
+/// Before insertion, `expand-abbrev' is executed if the inserted character does
+/// not have word syntax and the previous character in the buffer does.
+/// After insertion, `internal-auto-fill' is called if
+/// `auto-fill-function' is non-nil and if the `auto-fill-chars' table has
+/// a non-nil value for the inserted character.  At the end, it runs
+/// `post-self-insert-hook'.
+#[lisp_fn(intspec = "p")]
+pub fn self_insert_command(n: EmacsInt) -> () {
+    if n < 0 {
+        error!("Negative repetition argument {}", n);
+    } else if n < 2 {
+        call_raw!(Qundo_auto_amalgamate);
+    }
+
+    unsafe {
+        // Barf if the key that invoked this was not a character.
+        let c = LispObject::from_raw(globals.f_last_command_event)
+            .as_character()
+            .unwrap_or_else(|| bitch_at_user());
+
+        let character = translate_char(globals.f_Vtranslation_table_for_input, c as libc::c_int);
+        let val = internal_self_insert(character, n);
+        if val == 2 {
+            Fset(Qundo_auto__this_command_amalgamating, Qnil);
+            frame_make_pointer_invisible(selected_frame().as_frame_or_error().as_mut());
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn keys_of_cmds() {
     let global_map = current_global_map().to_raw();
 
     unsafe {
-        let A = CString::new("beginning-of-line").unwrap();
-        initial_define_key(global_map, Ctl('A'), A.as_ptr());
-        let B = CString::new("backward-char").unwrap();
-        initial_define_key(global_map, Ctl('B'), B.as_ptr());
-        let E = CString::new("end-of-line").unwrap();
-        initial_define_key(global_map, Ctl('E'), E.as_ptr());
-        let F = CString::new("forward-char").unwrap();
-        initial_define_key(global_map, Ctl('F'), F.as_ptr());
+        let sic = CString::new("self-insert-command").unwrap().as_ptr();
+        initial_define_key(global_map, Ctl('I'), sic);
+        for n in 0x20..0x7e {
+            initial_define_key(global_map, n, sic);
+        }
+        for n in 0xa0..0xff {
+            initial_define_key(global_map, n, sic);
+        }
+
+        let bol = CString::new("beginning-of-line").unwrap().as_ptr();
+        initial_define_key(global_map, Ctl('A'), bol);
+        let bc = CString::new("backward-char").unwrap().as_ptr();
+        initial_define_key(global_map, Ctl('B'), bc);
+        let eol = CString::new("end-of-line").unwrap().as_ptr();
+        initial_define_key(global_map, Ctl('E'), eol);
+        let fc = CString::new("forward-char").unwrap().as_ptr();
+        initial_define_key(global_map, Ctl('F'), fc);
     }
 }
 
diff --git a/rust_src/src/eval_macros.rs b/rust_src/src/eval_macros.rs
index 09c7fb1..6168270 100644
--- a/rust_src/src/eval_macros.rs
+++ b/rust_src/src/eval_macros.rs
@@ -55,8 +55,7 @@ macro_rules! call_raw {
     ($func:expr) => {{
         #[allow(unused_unsafe)]
         unsafe {
-            LispObject::from_raw(::remacs_sys::Ffuncall(1, &mut $func)
-            )
+            LispObject::from_raw(::remacs_sys::Ffuncall(1, &mut $func))
         }
     }}
 }
diff --git a/rust_src/src/lib.rs b/rust_src/src/lib.rs
index f4802dc..2ea1d1a 100644
--- a/rust_src/src/lib.rs
+++ b/rust_src/src/lib.rs
@@ -91,11 +91,6 @@ use alloc_unexecmacosx::OsxUnexecAlloc;
 #[global_allocator]
 static ALLOCATOR: OsxUnexecAlloc = OsxUnexecAlloc;
 
-#[no_mangle]
-pub extern "C" fn rust_initial_keys() {
-    cmds::initial_keys();
-}
-
 include!(concat!(env!("OUT_DIR"), "/c_exports.rs"));
 
 #[cfg(test)]
diff --git a/rust_src/src/lisp.rs b/rust_src/src/lisp.rs
index a8fc506..f3a70ca 100755
--- a/rust_src/src/lisp.rs
+++ b/rust_src/src/lisp.rs
@@ -1503,18 +1503,26 @@ impl LispObject {
 
     /// Nonzero iff X is a character.
     pub fn is_character(self) -> bool {
-        self.as_fixnum()
-            .map_or(false, |i| 0 <= i && i <= EmacsInt::from(MAX_CHAR))
+        self.as_character().is_some()
+    }
+
+    pub fn as_character(self) -> Option<Codepoint> {
+        match self.as_fixnum() {
+            None => None,
+            Some(i) => if 0 <= i && i <= EmacsInt::from(MAX_CHAR) {
+                Some(i as Codepoint)
+            } else {
+                None
+            },
+        }
     }
 
     /// Check if Lisp object is a character or not and return the codepoint
     /// Similar to CHECK_CHARACTER
     #[inline]
     pub fn as_character_or_error(self) -> Codepoint {
-        if !self.is_character() {
-            wrong_type!(Qcharacterp, self)
-        }
-        self.as_fixnum().unwrap() as Codepoint
+        self.as_character()
+            .unwrap_or_else(|| wrong_type!(Qcharacterp, self))
     }
 
     #[inline]
diff --git a/src/cmds.c b/src/cmds.c
index c9c5c23..8cd5af0 100644
--- a/src/cmds.c
+++ b/src/cmds.c
@@ -29,44 +29,6 @@ along with GNU Emacs.  If not, see <https://www.gnu.org/licenses/>.  */
 #include "keymap.h"
 #include "frame.h"
 
-static int internal_self_insert (int, EMACS_INT);
-
-/* Note that there's code in command_loop_1 which typically avoids
-   calling this.  */
-DEFUN ("self-insert-command", Fself_insert_command, Sself_insert_command, 1, 1, "p",
-       doc: /* Insert the character you type.
-Whichever character you type to run this command is inserted.
-The numeric prefix argument N says how many times to repeat the insertion.
-Before insertion, `expand-abbrev' is executed if the inserted character does
-not have word syntax and the previous character in the buffer does.
-After insertion, `internal-auto-fill' is called if
-`auto-fill-function' is non-nil and if the `auto-fill-chars' table has
-a non-nil value for the inserted character.  At the end, it runs
-`post-self-insert-hook'.  */)
-  (Lisp_Object n)
-{
-  CHECK_NUMBER (n);
-
-  if (XINT (n) < 0)
-    error ("Negative repetition argument %"pI"d", XINT (n));
-
-  if (XFASTINT (n) < 2)
-    call0 (Qundo_auto_amalgamate);
-
-  /* Barf if the key that invoked this was not a character.  */
-  if (!CHARACTERP (last_command_event))
-    bitch_at_user ();
-  else {
-    int character = translate_char (Vtranslation_table_for_input,
-				    XINT (last_command_event));
-    int val = internal_self_insert (character, XFASTINT (n));
-    if (val == 2)
-      Fset (Qundo_auto__this_command_amalgamating, Qnil);
-    frame_make_pointer_invisible (SELECTED_FRAME ());
-  }
-
-  return Qnil;
-}
 
 /* Insert N times character C
 
@@ -74,7 +36,7 @@ a non-nil value for the inserted character.  At the end, it runs
    return 0.  A value of 1 indicates this *might* not have been simple.
    A value of 2 means this did things that call for an undo boundary.  */
 
-static int
+int
 internal_self_insert (int c, EMACS_INT n)
 {
   int hairy = 0;
@@ -284,18 +246,4 @@ syms_of_cmds (void)
 	       doc: /* Hook run at the end of `self-insert-command'.
 This is run after inserting the character.  */);
   Vpost_self_insert_hook = Qnil;
-
-  defsubr (&Sself_insert_command);
-}
-
-void
-keys_of_cmds (void)
-{
-  int n;
-
-  initial_define_key (global_map, Ctl ('I'), "self-insert-command");
-  for (n = 040; n < 0177; n++)
-    initial_define_key (global_map, n, "self-insert-command");
-  for (n = 0240; n < 0400; n++)
-    initial_define_key (global_map, n, "self-insert-command");
 }
diff --git a/src/emacs.c b/src/emacs.c
index a3860ca..b4f474b 100644
--- a/src/emacs.c
+++ b/src/emacs.c
@@ -1568,7 +1568,6 @@ Using an Emacs configured with --with-x-toolkit=lucid does not have this problem
       keys_of_keyboard ();
       keys_of_keymap ();
       keys_of_window ();
-      rust_initial_keys ();
     }
   else
     {
diff --git a/src/lisp.h b/src/lisp.h
index 588ee09..02fbc15 100644
--- a/src/lisp.h
+++ b/src/lisp.h
@@ -3411,6 +3411,7 @@ aset_multibyte_string(register Lisp_Object array, EMACS_INT idxval, int c);
 /* Defined in cmds.c */
 extern void syms_of_cmds (void);
 extern void keys_of_cmds (void);
+extern int internal_self_insert (int, EMACS_INT);
 
 /* Defined in coding.c.  */
 extern Lisp_Object detect_coding_system (const unsigned char *, ptrdiff_t,
```

File: rust_src/remacs-sys/lib.rs [hunk 1]
```diff
@@ -1798,3 +1798,10 @@ fn basic_size_and_align() {
     assert!(::std::mem::size_of::<ptrdiff_t>() == 8);
     assert!(::std::isize::MAX == 9223372036854775807);
 }
+
+extern "C" {
+    pub fn internal_self_insert(c: EmacsInt, n: EmacsInt) -> EmacsInt;
+    pub fn frame_make_pointer_invisible(frame: *mut Lisp_Frame);
+    pub fn bitch_at_user();
```
Comment (Wilfred) [inline rust_src/remacs-sys/lib.rs]: Does this function return? If not, we should use `!` in its type.

Comment (shaleh) [reply rust_src/remacs-sys/lib.rs]: Yes, it should have a `!`.

**Pull request merged**